peps/pep-0817.rst

```diff
@@ -0,0 +1,2528 @@
+resorting to source builds. Variant wheels can be used to publish
+variants of vLLM built against different PyTorch versions, therefore
+enabling upstream to easily provide support for multiple versions
+simultaneously.
+
+The optional ``abi_dependency`` extension can be used to build multiple
+``vllm`` variants that are pinned to different PyTorch versions, e.g.:
+
+- ``vllm-0.11.0-...-torch29.wheel`` with
+  ``abi_dependency :: torch :: 2.9``
+- ``vllm-0.11.0-...-torch28.wheel`` with
+  ``abi_dependency :: torch :: 2.8``
+- ``vllm-0.11.0-...-torch27.wheel`` with
+  ``abi_dependency :: torch :: 2.7``
+
+
+Security implications
+=====================
+
+The proposal introduces a plugin system for querying the system
+capabilities in order to determine variant wheel capability. The system
+permits specifying additional Python packages providing the plugins
+in the package index metadata. Installers and other tools that need
+to determine whether a particular wheel is installable, or select
+the most preferred variant among multiple variant wheels, may need
+to install these packages and execute the code within them while
+resolving dependencies or processing wheels.
+
+This elevates the supply-chain attack potential by introducing two new
+points for malicious actors to inject arbitrary code payload:
+
+1. Publishing a version of a variant provider plugin or one of its
+   dependencies with malicious code.
+
+2. Introducing a malicious variant provider plugin in an existing
+   package metadata.
+
+While such attacks are already possible at the package dependency level,
+it needs to be emphasized that in some scenarios the affected tools are
+executed with elevated privileges, e.g. when installing packages for
+multi-user systems, while the installed packages are only used with
+regular user privileges afterwards. Therefore, variant provider plugins
+could introduce a Remote Code Execution vulnerability with elevated
+privileges.
+
+A similar issue already exists in the packaging ecosystem when packages
+are installed from source distributions, whereas build backends
+and other build dependencies are installed and executed. However,
+various tools operating purely on wheels, as well as users using
+tool-specific options to disable use of source distributions,
+have been relying on the assumption that no code external to the system
+will be executed while resolving dependencies, installing a wheel or
+otherwise processing it. To uphold this assumption, the proposal
+explicitly requires that provider plugin packages are never installed
+without explicit user consent.
+
+The `Overview`_ section of the specification provides further
+suggestions that aim to improve both security and the user experience.
+It is expected that a limited subset of popular provider plugins will
+be either vendored by the installer, eliminating the use of packages
+external to the tool altogether, or pinned to specific versions,
```

Comment:
The referenced section (Specification -> Overview) notes that pinning is to be avoided. Here, it is advised.